// epic-free-games/dist/modules/epic-free-games/epic-free-games.util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EpicFreeGamesUtil = void 0;
var deepmerge_1 = __importDefault(require("@fastify/deepmerge"));
var dayjs_1 = __importDefault(require("dayjs"));
var deepmerge = (0, deepmerge_1.default)();
var EpicFreeGamesUtil = /** @class */ (function () {
    function EpicFreeGamesUtil(options) {
        this.options = {
            country: 'US',
            includeAll: false,
            locale: 'en-US'
        };
        this.options = deepmerge(this.options, options);
    }
    EpicFreeGamesUtil.prototype.getSafeCountry = function (country) {
        return typeof country === 'string' ? country : this.options.country || 'US';
    };
    EpicFreeGamesUtil.prototype.getSafeIncludeAll = function (includeAll) {
        return typeof includeAll === 'boolean' ? includeAll : this.options.includeAll || false;
    };
    EpicFreeGamesUtil.prototype.getSafeLocale = function (locale) {
        return typeof locale === 'string' ? locale : this.options.locale || 'en-US';
    };
    EpicFreeGamesUtil.prototype.isBaseGame = function (data) {
        var includeAll = this.getSafeIncludeAll(data.includeAll);
        return includeAll ? true : data.offerGame.offerType === 'BASE_GAME';
    };
    EpicFreeGamesUtil.prototype.hasPromotionalOffers = function (data) {
        var _a, _b;
        return ((_b = (_a = data.offerGame.promotions) === null || _a === void 0 ? void 0 : _a.promotionalOffers) === null || _b === void 0 ? void 0 : _b.length) !== 0;
    };
    EpicFreeGamesUtil.prototype.hasUpcomingPromotionalOffers = function (data) {
        var _a, _b;
        return ((_b = (_a = data.offerGame.promotions) === null || _a === void 0 ? void 0 : _a.upcomingPromotionalOffers) === null || _b === void 0 ? void 0 : _b.length) !== 0;
    };
    EpicFreeGamesUtil.prototype.isFree = function (data) {
        var _a, _b;
        return ((_b = (_a = data.offerGame.price) === null || _a === void 0 ? void 0 : _a.totalPrice) === null || _b === void 0 ? void 0 : _b.discountPrice) === 0;
    };
    EpicFreeGamesUtil.prototype.inThisWeek = function (data) {
        var _a, _b, _c, _d, _e, _f;
        return ((0, dayjs_1.default)().isAfter((_c = (_b = (_a = data.offerGame.promotions) === null || _a === void 0 ? void 0 : _a.promotionalOffers[0]) === null || _b === void 0 ? void 0 : _b.promotionalOffers[0]) === null || _c === void 0 ? void 0 : _c.startDate) &&
            (0, dayjs_1.default)().isBefore((_f = (_e = (_d = data.offerGame.promotions) === null || _d === void 0 ? void 0 : _d.promotionalOffers[0]) === null || _e === void 0 ? void 0 : _e.promotionalOffers[0]) === null || _f === void 0 ? void 0 : _f.endDate));
    };
    EpicFreeGamesUtil.prototype.inNextWeek = function (data) {
        var _a, _b, _c, _d, _e, _f;
        return ((0, dayjs_1.default)()
            .add(1, 'week')
            .isAfter((_c = (_b = (_a = data.offerGame.promotions) === null || _a === void 0 ? void 0 : _a.upcomingPromotionalOffers[0]) === null || _b === void 0 ? void 0 : _b.promotionalOffers[0]) === null || _c === void 0 ? void 0 : _c.startDate) &&
            (0, dayjs_1.default)()
                .add(1, 'week')
                .isBefore((_f = (_e = (_d = data.offerGame.promotions) === null || _d === void 0 ? void 0 : _d.upcomingPromotionalOffers[0]) === null || _e === void 0 ? void 0 : _e.promotionalOffers[0]) === null || _f === void 0 ? void 0 : _f.endDate));
    };
    EpicFreeGamesUtil.prototype.willBeFree = function (data) {
        var _a, _b, _c, _d;
        return (((_d = (_c = (_b = (_a = data.offerGame.promotions) === null || _a === void 0 ? void 0 : _a.upcomingPromotionalOffers[0]) === null || _b === void 0 ? void 0 : _b.promotionalOffers[0]) === null || _c === void 0 ? void 0 : _c.discountSetting) === null || _d === void 0 ? void 0 : _d.discountPercentage) === 0);
    };
    return EpicFreeGamesUtil;
}());
exports.EpicFreeGamesUtil = EpicFreeGamesUtil;
