// epic-free-games/dist/modules/epic-free-games/epic-free-games.module.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EpicFreeGames = void 0;
var epic_free_games_util_1 = require("./epic-free-games.util");
var epic_games_1 = require("../epic-games");
/**
 * @author Aykut Saki <aykutsakisocial@gmail.com>
 */
var EpicFreeGames = /** @class */ (function () {
    function EpicFreeGames(options) {
        this.epicFreeGamesUtil = new epic_free_games_util_1.EpicFreeGamesUtil(options);
        this.epicGames = new epic_games_1.EpicGames();
    }
    EpicFreeGames.prototype.getGames = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var country, locale, offerGames, currentGames, nextGames;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        country = this.epicFreeGamesUtil.getSafeCountry(options.country);
                        locale = this.epicFreeGamesUtil.getSafeLocale(options.locale);
                        return [4 /*yield*/, this.epicGames.getGames({ country: country, locale: locale })];
                    case 1:
                        offerGames = _a.sent();
                        currentGames = this.getCurrentGames({ offerGames: offerGames, includeAll: options.includeAll });
                        nextGames = this.getNextGames({ offerGames: offerGames, includeAll: options.includeAll });
                        return [2 /*return*/, { currentGames: currentGames, nextGames: nextGames }];
                }
            });
        });
    };
    EpicFreeGames.prototype.getCurrentGames = function (data) {
        var _this = this;
        return data.offerGames.filter(function (offerGame) {
            return _this.epicFreeGamesUtil.isBaseGame({ offerGame: offerGame, includeAll: data.includeAll }) &&
                _this.epicFreeGamesUtil.hasPromotionalOffers({ offerGame: offerGame }) &&
                _this.epicFreeGamesUtil.isFree({ offerGame: offerGame }) &&
                _this.epicFreeGamesUtil.inThisWeek({ offerGame: offerGame });
        });
    };
    EpicFreeGames.prototype.getNextGames = function (data) {
        var _this = this;
        return data.offerGames.filter(function (offerGame) {
            return _this.epicFreeGamesUtil.isBaseGame({ offerGame: offerGame, includeAll: data.includeAll }) &&
                _this.epicFreeGamesUtil.hasUpcomingPromotionalOffers({ offerGame: offerGame }) &&
                _this.epicFreeGamesUtil.willBeFree({ offerGame: offerGame }) &&
                _this.epicFreeGamesUtil.inNextWeek({ offerGame: offerGame });
        });
    };
    return EpicFreeGames;
}());
exports.EpicFreeGames = EpicFreeGames;
